static/src/js/zustand/slices/createGranulesSlice.ts

```diff
@@ -0,0 +1,193 @@
+
+const initialState = {
+  collectionConceptId: null,
+  count: null,
+  isLoaded: false,
+  isLoading: false,
+  loadTime: 0,
+  items: []
+}
+
+// Cancel token to cancel pending search requests
+const granuleSearchCancelTokens: Record<string, CancelTokenSource | undefined> = {}
+
+const createGranulesSlice: ImmerStateCreator<GranulesSlice> = (set, get) => ({
+  granules: {
+    granules: initialState,
+
+    getGranules: async () => {
+      const {
+        dispatch: reduxDispatch,
+        getState: reduxGetState
+      } = configureStore()
+      const reduxState = reduxGetState()
+
+      const currentState = get()
+      const earthdataEnvironment = getEarthdataEnvironment(currentState)
+      const collectionId = getCollectionId(currentState)
+      const collectionMetadata = getFocusedCollectionMetadata(currentState)
+      const collections = getCollectionsById(currentState)
+      const collectionById = collections[collectionId]
+
+      const { authToken } = reduxState
+
+      // Extract granule search parameters from redux specific to the focused collection
+      const extractedGranuleParams = extractGranuleSearchParams(reduxState, collectionId)
+
+      // Format and structure data from Redux to be sent to CMR
+      const granuleParams = prepareGranuleParams(
+        // Use focusedCollection and collection search results together to build the granule
+        // search parameters. This will ensure isOpenSearch can be properly determined before
+        // retrieving full collection metadata
+        {
+          ...collectionMetadata,
+          ...collectionById
+        },
+        extractedGranuleParams
+      )
+
+      // If cancel token is set, cancel the previous request(s)
+      if (granuleSearchCancelTokens[collectionId]) {
+        granuleSearchCancelTokens[collectionId].cancel()
+        granuleSearchCancelTokens[collectionId] = undefined
+      }
+
+      const {
+        isOpenSearch,
+        pageNum
+      } = granuleParams
+
+      if (pageNum === 1) {
+        // If the collection ID matches the collectionConceptId already saved, and the collection is
+        // not OpenSearch, don't fetch granules.
+        // In `getCollectionMetadata` we call `getGranules` as soon as we can to fetch CMR granules right away.
+        // But that request will be empty for OpenSearch collections, which aren't stored in CMR.
+        // So after the collection metadata is returned we call `getGranules` again. This results
+        // in a double fetch for OpenSearch Granules, but the two requests are made to different
+        // endpoints. This `if` ensures we don't double fetch CMR granules.
+        // TODO This is causing a double fetch for OpenSearch granules (both to the same OpenSearch endpoint)
+        // TODO when loaded from the search results. The first request gets cancelled, but it would be nice to avoid that
+        if (
+          collectionId === currentState.granules.granules.collectionConceptId
```

Comment:
This logic is keeping granule searches from happening when the query changes, like adding temporal or spatial values from the granules page